### PR TITLE
add simple tests to TV denoise implementation

### PR DESCRIPTION
**Pull request by tjlane**
This PR is effectively a re-write of all the code needed to do single-pass TV denoising. It is a mega-PR, also implementing CI (testing, lint/formatting, type-checking).

Functionality not represented:
- dataset-to-dataset scaling
- iterative TV

### CODE REVIEW

File: test/unit/test_tv.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        },
+        spacegroup=sg_1,
+        cell=cell,
+    ).infer_mtz_dtypes()
+    ds.set_index(["H", "K", "L"], inplace=True)
+
+    return ds
+
+
+def test_tv_denoise_difference_map_smoke(flat_difference_map: rs.DataSet) -> None:
```
Comment (alisiafadini) [inline test/unit/test_tv.py:39]: What is smoke?

File: meteor/tv.py [hunk 2]
```diff
@@ -34,52 +57,36 @@ def tv_denoise_difference_map(
...
-        amplitude_label=TV_AMPLITUDE_LABEL,
-        phase_label=TV_PHASE_LABEL,
-    )
+    # TODO: verify correctness
+    final_map_coefficients = (np.fft.fftn(final_map_array)[::TV_MAP_SAMPLING, ::TV_MAP_SAMPLING, ::TV_MAP_SAMPLING]).flatten()
+    print(final_map_coefficients.shape, len(difference_map_coefficients))
 
-    # ^^^ replace this with something better!
+    # TODO: need to be sure HKLs line up
+    difference_map_coefficients[[TV_AMPLITUDE_LABEL]] = np.abs(final_map_coefficients)
```
Comment (alisiafadini) [inline meteor/tv.py:86]: we should make this its own function

File: test/unit/test_tv.py (hunk 1 above)
Comment (tjlane) [reply test/unit/test_tv.py:39]: [smoke tests](https://en.wikipedia.org/wiki/Smoke_testing_(software)) usually just make sure a function runs and returns the intended kind of output, without ensuring that output is correct

File: meteor/tv.py (hunk 2 above)
Comment (tjlane) [reply meteor/tv.py:86]: absolut

File: meteor/tv.py [hunk 3]
```diff
@@ -33,53 +52,52 @@ def tv_denoise_difference_map(
...
-            weight=tv_lambda,
-            max_num_iter=TV_MAX_NUM_ITER,
+        denoised_map = _tv_denoise_array(
+            map_as_array=difference_map_as_array, weight=tv_lambda
+        )
+        return negentropy(denoised_map.flatten())
+
+    optimal_lambda: float
+    if lambda_values_to_scan is not None:
+        highest_negentropy = -1e8
```
Comment (alisiafadini) [inline meteor/tv.py:65]: add comment on this value?

File: meteor/tv.py [hunk 4]
```diff
@@ -33,53 +52,52 @@ def tv_denoise_difference_map(
...
+    _, dmin = resolution_limits(difference_map_coefficients)
+    final_map_coefficients = compute_coefficients_from_map(
+        ccp4_map=final_map_as_ccp4,
+        high_resolution_limit=dmin,
+        amplitude_label=difference_map_amplitude_column,
+        phase_label=difference_map_phase_column,
     )
 
-    # ^^^ replace this with something better!
+    # ducktype for now
```
Comment (alisiafadini) [inline meteor/tv.py:97]: FIXITT

File: test/conftest.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        spacegroup=sg_1,
+        cell=cell,
+    ).infer_mtz_dtypes()
+    ds.set_index(["H", "K", "L"], inplace=True)
+
+    return ds
+
+
+@fixture
+def flat_difference_map() -> rs.DataSet:
```
Comment (alisiafadini) [inline test/conftest.py:36]: was ist flat?

File: test/conftest.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+from meteor.utils import canonicalize_amplitudes
+
+
+@fixture
+def random_intensities() -> rs.DataSet:
+    """
+    A simple 10x10x10 P1 dataset, with random intensities
+    """
+
+    params = (10.0, 10.0, 10.0, 90.0, 90.0, 90.0)
```
Comment (alisiafadini) [inline test/conftest.py:14]: This is a lot of copypasta, please reduce

File: meteor/tv.py (hunk 4 above)
Comment (tjlane) [reply meteor/tv.py]: implemented an optionally-returned object that we can use to hold all the output you might want, check it out

File: test/conftest.py (hunk 6 above)
Comment (tjlane) [reply test/conftest.py]: good point, done

File: test/unit/test_tv.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    # testmap = compute_map_from_coefficients(
+    #     map_coefficients=denoised_map,
+    #     amplitude_label=TEST_AMPLITUDE_LABEL,
+    #     phase_label=TEST_PHASE_LABEL,
+    #     map_sampling=1
+    # )
+    # testmap.write_ccp4_map("denoised.ccp4")
+
+
+def test_tv_denoise_difference_map_specific_lambdas(noise_free_map: rs.DataSet, noisy_map: rs.DataSet) -> None:
```
[the remark this answers is not in the document]
Comment (tjlane) [reply test/unit/test_tv.py]: done

File: meteor/utils.py [hunk 8]
```diff
@@ -1,77 +1,88 @@
...
 
-    """
+def canonicalize_amplitudes(
+    dataset: rs.DataSet,
+    *,
+    amplitude_label: str,
+    phase_label: str,
+    inplace: bool = False,
+) -> rs.DataSet | None:
+    dataset.canonicalize_phases(inplace=inplace)
```
[the remark this answers is not in the document]
Comment (tjlane) [reply meteor/utils.py]: fixed

File: test/unit/test_utils.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+@pytest.mark.parametrize("inplace", [False, True])
+def test_canonicalize_amplitudes(
+    inplace: bool, flat_difference_map: rs.DataSet
+) -> None:
+    amplitude_label = "DF"
+    phase_label = "PHIC"
+
+    if inplace:
+        canonicalized = flat_difference_map.copy(deep=True)
```
[the remark this answers is not in the document]
Comment (tjlane) [reply test/unit/test_utils.py]: done

File: meteor/tv.py [hunk 10]
```diff
@@ -1,85 +1,189 @@
...
+    compute_coefficients_from_map,
+    compute_map_from_coefficients,
+    numpy_array_to_map,
+    resolution_limits,
+)
+from .validate import negentropy
+
+
+@dataclass
+class TvDenoiseResult:
```
Comment (alisiafadini) [inline meteor/tv.py:25]: add a map_sampling attribute

File: meteor/tv.py [hunk 11]
```diff
@@ -33,53 +52,52 @@ def tv_denoise_difference_map(
...
-    denoised_dataset = compute_coefficients_from_map(
-        ccp4_map=ccp4_map,
-        high_resolution_limit=high_resolution_limit,
-        amplitude_label=TV_AMPLITUDE_LABEL,
-        phase_label=TV_PHASE_LABEL,
+    _, dmin = resolution_limits(difference_map_coefficients)
+    final_map_coefficients = compute_coefficients_from_map(
+        ccp4_map=final_map_as_ccp4,
+        high_resolution_limit=dmin,
+        amplitude_label=difference_map_amplitude_column,
```
[the remark this answers is not in the document]
Comment (tjlane) [reply meteor/tv.py:175]: let's discuss this one, I think I disagree

File: test/unit/test_tv.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+def noise_free_map() -> rs.DataSet:
+    return displaced_single_atom_difference_map_coefficients(noise_sigma=0.0)
+
+
+@fixture
+def noisy_map() -> rs.DataSet:
+    return displaced_single_atom_difference_map_coefficients(noise_sigma=0.03)
+
+
+def test_tv_denoise_difference_map_smoke(flat_difference_map: rs.DataSet) -> None:
```
[the remark this answers is not in the document]
Comment (tjlane) [reply test/unit/test_tv.py]: yep done

File: test/unit/test_utils.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+def test_cut_resolution(
+    random_intensities: rs.DataSet, dmax_limit: float, dmin_limit: float
+) -> None:
+    dmax_before_cut, dmin_before_cut = utils.resolution_limits(random_intensities)
+    if not dmax_limit:
+        expected_max_dmax = dmax_before_cut
+    else:
+        expected_max_dmax = dmax_limit
+
+    if not dmin_limit:
```
[the remark this answers is not in the document]
Comment (tjlane) [reply test/unit/test_utils.py]: Attempted. See what you think.

**Comment by tjlane** [comment]
Pending a final review, I think we should merge this and then keep going.

To do after merge: 
- much stronger testing of implemented TV code
- think more carefully about sampling during TV denoising
- battle test, time golden section method

**Pull request merged**